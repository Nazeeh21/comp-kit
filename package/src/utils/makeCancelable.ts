export const makeCancelable = <T>(
  promise: Promise<T>
): { promise: Promise<T>; cancel: () => void } => {
  let hasCanceled = false;

  const wrappedPromise = new Promise<T>((resolve, reject) => {
    promise.then(
      value => {
        if (!hasCanceled) {
          resolve(value);
        }
      },
      error => {
        if (!hasCanceled) {
          reject(error);
        }
      }
    );
  });

  return {
    promise: wrappedPromise,
    cancel: () => {
      hasCanceled = true;
    },
  };
};
