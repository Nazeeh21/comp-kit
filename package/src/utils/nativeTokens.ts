import { useEffect, useState } from 'react';
import { Chain } from 'viem';
import { bsc, mainnet, polygon, polygonMumbai } from 'viem/chains';

const nativeTokens: Record<string, string> = {
  [mainnet.id]: 'ETH',
  [bsc.id]: 'BNB',
  [polygon.id]: 'MATIC',
  [polygonMumbai.id]: 'MATIC',
};

export function useGetNativeToken({ chain }: { chain: Chain }) {
  const [nativeToken, setNativeToken] = useState('ETH');
  useEffect(() => {
    if (nativeTokens[chain.id]) {
      setNativeToken(nativeTokens[chain.id]);
    }
  }, [chain]);
  return {
    nativeToken,
  };
}
