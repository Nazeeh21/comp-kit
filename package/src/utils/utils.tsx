import { Address } from 'viem';
import * as chains from 'viem/chains';
import { requestObject } from '../components/test/utils';

declare global {
  interface Window {
    ethereum: {
      off(arg0: string, handleChainChanged: (chainId: string) => void): unknown;
      on(arg0: string, arg1: (chainId: string) => void): unknown;
      enable(): unknown;
      request: typeof requestObject;
    };
  }
}

/**
 * Gets the chain object for the given chain id.
 * @param chainId - Chain id of the target EVM chain.
 * @returns Viem's chain object.
 */
export function getChain(chainId: number): chains.Chain | undefined {
  try {
    for (const chain of Object.values(chains)) {
      if (chain.id === chainId) {
        return chain;
      }
      if (chainId === 0) return chains.mainnet;
    }
  } catch (error) {
    // this error indicates that the chain user has switched to is not supported by viem
    // TODO: handle this error using native error component
    throw new Error(`Chain with id ${chainId} not found in viem`);
  }
}

export function storePrevWallet(wallet: 'MetaMask' | 'WalletConnect') {
  localStorage.setItem('prevWallet', wallet);
}

export function getPrevWallet() {
  return localStorage.getItem('prevWallet');
}

export function removePrevWallet() {
  localStorage.removeItem('prevWallet');
}

interface PrevAccount {
  accounts: Address[];
  chain: chains.Chain;
}

export function storePrevAccount({ accounts, chain }: PrevAccount) {
  localStorage.setItem(
    'prevAccount',
    JSON.stringify({
      data: { accounts, chain },
    })
  );
}

export function getPrevAccount() {
  const prevAccount = localStorage.getItem('prevAccount');
  if (!prevAccount) return undefined;
  return JSON.parse(prevAccount) as { data: PrevAccount };
}

export function removePrevAccount() {
  localStorage.removeItem('prevAccount');
}
