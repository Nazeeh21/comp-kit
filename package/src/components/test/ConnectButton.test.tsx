import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { mainnet } from 'viem/chains';
import { ConnectButton, KitProvider } from '../../index';
import { requestObject } from './utils';

declare global {
  interface Window {
    ethereum: { request: typeof requestObject };
  }
}

describe('test ConnectButton', () => {
  const Comp: React.FC = () => (
    <KitProvider initialChain={mainnet} supportedChains={[mainnet]}>
      <ConnectButton />
    </KitProvider>
  );
  xit('renders ConnectButton', async () => {
    let container: unknown;
    await act(() => {
      container = renderer.create(<Comp />).toJSON();
    });
    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
