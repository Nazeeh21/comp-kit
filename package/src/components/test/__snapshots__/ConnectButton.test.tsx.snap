// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ConnectButton renders ConnectButton 1`] = `
<button
  className="c-gJLZyQ c-gJLZyQ-hvqSqD-color-light"
  onClick={[Function]}
>
  Connect Wallet
</button>
`;
